stan/math/rev/fun/add.hpp

```diff
@@ -0,0 +1,538 @@
+ * the varis for A and B if A or B is a var. It
+ * also instatiates and stores pointers to varis
+ * for all elements of A + B.
+ *
+ * @tparam Ta type of elements in matrix A
+ * @tparam R number of rows (common to A and B)
+ * @tparam C number of columns (common to A and B)
+ */
+
+template <typename Ta, int R, int C>
+class add_mat_vari<Ta, double, R, C> : public vari {
+ public:
+  int rows_;
+  int cols_;
+  int size_;
+  double* Ad_;
+  double* Bd_;
+  vari** variRefA_;
+  vari** variRefAplusB_;
+
+  /**
+   * Constructor for add_mat_vari.
+   *
+   * All memeory is allocated in ChainableStack's
+   * stack_alloc arena.
+   *
+   * It is critical for the efficiency of this object
+   * that the constructor create new varis that aren't
+   * popped onto the var_stack_, but rather are
+   * popped onto the var_nochain_stack_. This is
+   * controlled by the second argument to
+   * vari's constructor.
+   *
+   * @param A matrix
+   * @param  B matrix
+   **/
+  add_mat_vari(const Eigen::Matrix<Ta, R, C>& A,
+               const Eigen::Matrix<double, R, C>& B)
+      : vari(0.0),
+        rows_(A.rows()),
+        cols_(B.cols()),
+        size_(A.size()),
+        Ad_(ChainableStack::instance_->memalloc_.alloc_array<double>(size_)),
+        Bd_(ChainableStack::instance_->memalloc_.alloc_array<double>(size_)),
+        variRefA_(
+            ChainableStack::instance_->memalloc_.alloc_array<vari*>(size_)),
+        variRefAplusB_(
+            ChainableStack::instance_->memalloc_.alloc_array<vari*>(size_)) {
+    using Eigen::Map;
+    Map<matrix_vi>(variRefA_, rows_, cols_) = A.vi();
+    Map<matrix_d> Ad(Ad_, rows_, cols_);
+    Map<matrix_d> Bd(Bd_, rows_, cols_);
+    Ad = A.val();
+    Bd = B;
+    Map<matrix_vi>(variRefAplusB_, rows_, cols_)
+        = (Ad + Bd).unaryExpr([](double x) { return new vari(x, false); });
+  }
+
+  virtual void chain() {
+    using Eigen::Map;
+    matrix_d adjAplusB(rows_, cols_);
+    adjAplusB = Map<matrix_vi>(variRefAplusB_, rows_, cols_).adj();
+    Map<matrix_vi>(variRefA_, rows_, cols_).adj() += adjAplusB;
+  }
+};
+
+/**
+ * Return the sum of two matrices
+ * @tparam Mat1 type of first matrix
+ * @tparam Mat2 type of second matrix
+ *
+ * @param[in] A matrix
+ * @param[in] B matrix
+ * @return Sum of two matrices
+ */
+template <typename Mat1, typename Mat2,
+          require_all_eigen_t<Mat1, Mat2>* = nullptr,
+          require_any_eigen_vt<is_var, Mat1, Mat2>* = nullptr>
+auto add(const Mat1& A, const Mat2& B) {
+  using Ta = value_type_t<Mat1>;
+  using Tb = value_type_t<Mat2>;
+  constexpr int R = Mat1::RowsAtCompileTime;
+  constexpr int C = Mat1::ColsAtCompileTime;
+  check_matching_dims("add", "A", A, "B", B);
+  check_not_nan("add", "A", A);
+  check_not_nan("add", "B", B);
+
+  // Manage memory in arena allocator
+  add_mat_vari<Ta, Tb, R, C>* baseVari = new add_mat_vari<Ta, Tb, R, C>(A, B);
+  Eigen::Matrix<var, R, C> AplusB_v(A.rows(), A.cols());
```

Comment:
I'm realising that this is a broader issue with the generalised Eigen rev specialisations, that they can take any kind of input, but always return a ```Matrix<var, R, C>``` as the output.

@t4c1 What are your thoughts here?

Reply:
That is true. Rev (anfd fwd) expressions would be quite complicated to implement with current rev matrix being implemented as array of structs. We have some plans for struct of arrays implementations in #1805. That will allow to return two eigen expressions without doing custom Eigen expressions.

Reply:
As far as I understand, it shouldn't be that complicated. As long as ```add_mat_vari``` is generalised to work with ```Matrix```/```Array``` inputs, the only difficulty is whether the return should be a ```Matrix``` or ```Array```, since the autodiff implementations won't be returning expressions.

We could use something like:
```
template<typename EigT, int R, int C, typename Enable = void>
struct eigen_return_type {};

template<typename EigT, int R, int C>
struct eigen_return_type<EigT, R, C, require_t<is_eigen_matrix<plain_type_t<EigT>>>> {
  using type = Eigen::Matrix<value_type_t<EigT>, R, C>;
}

template<typename EigT, int R, int C>
struct eigen_return_type<EigT, R, C, require_t<is_eigen_array<plain_type_t<EigT>>>> {
  using type = Eigen::Array<value_type_t<EigT>, R, C>;
}

template<typename EigT, int R, int C>
using eigen_return_t = typename eigen_return_type<EigT, R, C>::type;
```

Then in the function, the returned object can be declared as:
```
eigen_return_t<Mat1, R, C> AplusB_v(A.rows(), A.cols());

AplusB_v.matrix().vi()
    = Eigen::Map<matrix_vi>(&baseVari->variRefAplusB_[0], A.rows(), A.cols());
```

Reply:
We also need to look at restricting the current templates to working with the same Eigen types, since the current templating allows for ```multiply(Matrix, Array)```, which I don't think should be allowed

Reply:
@dpsimpson Sorry for derailing the pull here! I think you can just ignore this conversation and continue as normal. I'll open up a separate issue and make any changes after we settle on a solution.

Reply:
What you suggested solves the problem of deciding whether to return `Matrix` or `Array`. However that is not the same as returning an expresion.

Reply:
But are these methods ever going to be able to return an expression? Don't they need to instantiate some container of ```var``` with the calculated ```vari(s)``` and then return that?

Reply:
I have to say I don't exactely have deep understanding of autodiff, but I believe with some work it is possible.

Reply:
Given that the ```varis``` are just a contiguous section of memory, if an expression is returned it has to be done to a particular 'shape' and plain type.

It could be something like:
```
Eigen::Map<matrix_vi> VariRef(&baseVari->variRefAplusB_[0], A.rows(), A.cols());

return VariRef.unaryExpr([](vari** x) { return var(x); });
```

But that still requires deducing whether the 'base type' should be ```Matrix``` or ```Array```

Reply:
That should work. But I think it would not be any faster than current implementation. I think we can do better. I don't have exact plan how yet. I will start thinking about this after I finish with generalizations of functions.

Reply:
Given that the benefits of returning expressions are that only the needed parts of a matrix/array get evaluated, I don't know if we'll be able to achieve that since the actual calculations for the values and adjoints are done on the entirety of the input matrices and the results copied into ```varis```. I don't think there's way to propogate expressions that only evaluate the needed part of the matrix into ```vari``` and then return those as ```var```, since the calculations with ```varis``` can't return expressions.